### DISCUSSION TITLE
More than 2 texture coordinates

### DISCUSSION THREAD

**Discussion by kpeeters** [google/filament#7340] [General]
I have a need to set 3 texture coordinates as attributes to my vertex buffer. However, `VertexAttribute` only seems to know about `UV0` and `UV1`. Is there any workaround for this (short of multiplexing the two `float2` into a `float4` of the `COLOR` attribute, or similar tricks)?

**Comment by romainguy** [comment]
You can use any of the custom attributes for this. The names are just convenience.

**Comment by kpeeters** [comment]
So there is no hidden assumption somewhere that `COLOR` is a `float4`? There are only two attributes listed as `float2`, and those are the texture coordinates which are already in use.

**Reply by romainguy** [reply]
All attributes but uv0 and uv1 are float4. They don't have to be on the CPU side but they are read as float4 in the shaders. uv0 and uv1 are packed together in a single float4 attribute.

**Reply by kpeeters** [reply]
So I should be able to setup the attribute for my 3rd float texture coordinate with

    builder.attribute(VertexAttribute::COLOR, bufferIndex, VertexBuffer::AttributeType::FLOAT2);

(note: `FLOAT2`) feed the data through with

    vertexBuffer->setBufferAt(*engine, bufferIndex, 
                                VertexBuffer::BufferDescriptor(data, numVertices*sizeof(math::float2))

(note: `float2`) and get it out in the fragment shader with

    vec2 uv2 = getColor().xy

? Or is that a misunderstanding of what you meant with "they don't have to be float2 on the CPU side"?

**Reply by romainguy** [reply]
You can use the color attribute or any of the CUSTOMx attributes.

**Reply by kpeeters** [reply]
The CUSTOMx attributes do not seem to be available in the fragment shader, correct? The documentation (in section 4.2.11) has an example in which `getCustom0()` is called inside the fragment shader, but `matc` complains that this function is not available there (even though I have `custom0` listed in the `requires` array).

**Reply by kpeeters** [reply]
Does the documentation have an example on how to get the `getColor0()` info into the fragment shader? (in other words, how to introduce a `varying` in the vertex shader and read that in the fragment shader?).

**Reply by romainguy** [reply]
You can see an example here with the "eyeDirection": https://github.com/google/filament/blob/main/filament/src/materials/skybox.mat
